code/mapSwitching.dm

```diff
@@ -486,8 +441,149 @@ var/global/datum/mapSwitchHandler/mapSwitcher
+/datum/map_vote_holder
+	var/list/client/vote_map = list() // a map of ckeys to (a map of map_names to the ckey's current vote)
+	var/voters = 0
 
+	disposing()
+		clear_votes()
+		..()
+
+	proc/clear_votes()
+		src.vote_map.len = 0
+
+	proc/count_votes()
+		var/list/results = list()
+		var/list/map_vote_count = list()
+		var/list/map_vote_report = list()
+		for(var/key in vote_map)
+			mapSwitcher.passiveVotes.Remove(key)
+			var/client_vote_map = vote_map[key]
+			for(var/map_name in client_vote_map)
+				if(client_vote_map[map_name])
+					if(map_name in map_vote_count)
+						map_vote_count[map_name] += MAPVOTE_ACTIVE_WEIGHT
+					else
+						map_vote_count[map_name] = MAPVOTE_ACTIVE_WEIGHT
+					if(map_name in map_vote_report)
+						map_vote_report[map_name] += key
+					else
+						map_vote_report[map_name] = list(key)
+
+		var/list/passive_votes = mapSwitcher.passiveVotes
+		for(var/key in passive_votes)
+			var/map_name = passive_votes[key]
+			if(map_name in map_vote_count)
+				map_vote_count[map_name] += MAPVOTE_PASSIVE_WEIGHT
+			else
+				map_vote_count[map_name] = MAPVOTE_PASSIVE_WEIGHT
+			if(map_name in map_vote_report)
+				map_vote_report[map_name] += key
+			else
+				map_vote_report[map_name] = list(key)
+
+		results["tally"] = map_vote_count
+		results["report"] = map_vote_report
+		return results
+
+	proc/setup_client_vote_map(var/client/C)
+		if(!C.ckey)
+			return
+		var/key = C.ckey
+		if(key in vote_map)
+			return
+		var/list/maps = list()
+		for(var/map in mapSwitcher.playerPickable)
+			maps[map] = 0
+		src.vote_map[key] = maps
+		voters++
+
+	proc/get_client_votes(var/client/C)
+		if(!C.ckey)
+			return
+		var/key = C.ckey
+		if(!(key in vote_map))
+			return
+		var/list/maps = list()
+		var/list/client_vote_map = vote_map[key]
+		for(var/map in client_vote_map)
+			if(client_vote_map[map] > 0)
+				maps += map
+		return maps
+
+	proc/toggle_vote(subaction, href_list)
+		var/client/C = locate(href_list["client"])
+		var/map_name = subaction
+		if(!(C.ckey in vote_map))
+			setup_client_vote_map(C)
+		var/list/client_vote_map = vote_map[C.ckey]
+		if(map_name in client_vote_map)
+			client_vote_map[map_name] = !client_vote_map[map_name]
+
+	proc/all_yes(subaction, href_list)
+		var/client/C = locate(href_list["client"])
+		if(!(C.ckey in vote_map))
+			setup_client_vote_map(C)
+		var/list/client_vote_map = vote_map[C.ckey]
+		for(var/map_name in client_vote_map)
+			client_vote_map[map_name] = 1
+
+	proc/all_no(subaction, href_list)
+		var/client/C = locate(href_list["client"])
+		if(!(C.ckey in vote_map))
+			setup_client_vote_map(C)
+		var/list/client_vote_map = vote_map[C.ckey]
+		for(var/map_name in client_vote_map)
+			client_vote_map[map_name] = 0
+
+	proc/special_vote(var/client/C,var/map_name)
+		if(!(C.ckey in vote_map))
+			setup_client_vote_map(C)
+		var/list/client_vote_map = vote_map[C.ckey]
+		client_vote_map[map_name] = 1
+
+	proc/voting_box(var/obj/voting_box/V,var/map_name)
+		var/vref = "\ref[src]"
```

Comment:
```suggestion
		var/vref = "\ref[V]"
```

I would probably just use `var/vref = V` here though?

Reply:
resolved in pending commit